core/drivers/imx_i2c.c

```diff
@@ -420,22 +425,87 @@ static TEE_Result get_va(paddr_t pa, vaddr_t *va)
 	return TEE_ERROR_GENERIC;
 }
 
-static TEE_Result i2c_init(void)
+#if defined(CFG_DT) && !defined(CFG_EXTERNAL_DTB_OVERLAY)
+static const char *const dt_i2c_match_table[] = {
+	"fsl,imx21-i2c",
+};
+
+static TEE_Result i2c_mapped(const char *i2c_match)
 {
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	void *fdt = get_dt();
+	size_t size = 0;
+	size_t i = 0;
+	int off = 0;
+
+	if (!fdt)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	for (i = 0; i < ARRAY_SIZE(i2c_bus); i++) {
+		off = fdt_node_offset_by_compatible(fdt, off, i2c_match);
+		if (off <= 0)
+			break;
+
+		if (!(_fdt_get_status(fdt, off) & DT_STATUS_OK_SEC)) {
+			EMSG("i2c%zu not enabled\n", i + 1);
```

Comment:
Are you sure you'll get the right I2C instance index with this loop? There is no strong guarantee on nodes ordering in generated DTBs.

Reply:
I dont disagree with that...we could check if there is perhaps an alias associated to the node (ie, of_alias_get_id) but there is not a guarantee either that there will be one so the issue will be just the same and will have added more code..... no matter what you do, there will be no guarantee unless all i2c buses are declared with their addresses sorted alphabetically (which is what the kernel has).